Lib/glyphsLib/classes.py

```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
             self.metrics.append(GSMetricValue(position=0))
         self.metrics[metricIndex] = GSMetricValue(position=value)
 
+    @property
+    def alignmentZones(self):
+        # If there are values that are parsed from file format 2 or which are
+        # explicitly set return those
+        if self._alignmentZones is not None:
+            return self._alignmentZones
+
+        if len(self.metrics) == 0:
+            return []
+
+        validZoneMetrics = (
+            "ascender",
+            "cap height",
+            "x-height",
+            "baseline",
+            "descender",
+        )
+
+        zones = []
+        for index, fontMetric in enumerate(self.font.metrics):
+            if fontMetric.type not in validZoneMetrics:
+                continue
+            metric = self.metrics[index]
+            zone = GSAlignmentZone(pos=metric.position, size=metric.overshoot)
+            zones.append(zone)
+        return zones
+
+    @alignmentZones.setter
+    def alignmentZones(self, entries):
```

Comment:
looks like alignmentZones is a read-only (https://docu.glyphsapp.com/#GSFontMaster.alignmentZones) so maybe we should not add a setter.
I agree with @simoncozens when he wrote

> We should store things as v3 format internally to the object but provide accessors to get at information in a v2 way.

Reply:
Some of the existing tests assign `alignmentZones` like this for constructing objects via code only (as opposed to actual conversion). Happy to drop the setter and tests.

Reply:
If one wants to _set_ these alignment zones using Glyphs.app v3 scripting API (where alignmentZones is read-only), one would one do that exactly?
Does glyphsLib already support this? If not, what do we need to support it?

Reply:
Going by what I found in the code. The primary use case in the library seems to be [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/builder/blue_values.py#L55) setting the alignmentZones when converting from UFO to glyphs. 

Also tests [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/classes_test.py#L755), [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/writer_test.py#L301), [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/builder/builder_test.py#L495) and [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/builder/builder_test.py#L504) — my assumption being that a new feature or refactor should not break existing tests and thus a setter needs to be supported.

Reply:
i'm ok to keep a setter for alignmentZones, but I wonder maybe it should store them internally as v3 metricValues.
Say you go from UFO => glyphs and want to save as format_version=3, those zones will not be written out in the generated .glyphs file.
I think we should have a method to convert from v2 (alignmentZones and stems) to v3 metricValues and call that when saving a font and format_version=3.

Reply:
but that setter only sets `self._alignmentZones`, which as far as I understand is _not_ written out to .glyphs source when format_version=3,

```python
        if writer.format_version == 2:
            writer.writeObjectKeyValue(self, "alignmentZones", "if_true")
```

Reply:
so I presume that roundtripped glyphs file of yours has format version 2.. but when Glyphs.app v3 opens it up it automatically populate the metric values.
That's what we are missing here, a way to upgrade these metrics from v2 to v3. I am not saying that we should do within this same PR, just that this is something we would like to support

Reply:
Ah, I see, didn't understand that's what's happening. Yes, correct. 

I'll have a look at that, either as part of this or separate. Follows the same obtuse logic of font metrics vs master metric 😄 Happy to adopt that if there is a more general v2 > v3 structure in place. That section of code you quoted, I suppose, is at the heart of what would need a more elegant and transparent file format version upgrade.

Reply:
_How_ does one actually write something out in format 3? Tried to find tests that verify those other `self.format == 3` written things (e.g. [here onwards](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/Lib/glyphsLib/classes.py#L4234)), but could only find [this one](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/tests/writer_test.py#L1190). Am I completely off track or is there pretty little actually for outputting/writing/testing v3? Would testing the output of the writer in `writer_tests.py` the right spot to implement checking for a yet-to-be-done implementation of writing those metrics back to v3 files?

Reply:
> but could only find [this one](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/tests/writer_test.py#L1190).

This looks like the way to do this. While writing out, check the format and what data you have and act accordingly.